File: QuickSort.py
# QuickSort
# Partition

def quick_sort(arr, p, r):
    if (p < r):
        q = partition(arr, p, r)
        quick_sort(arr, p, q - 1)
        quick_sort(arr, q + 1, r)

def partition(arr, p, r):
    key = arr[r]
    i = p - 1
    for j in range(p, r):
        if (arr[j] < key):
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[r] = arr[r], arr[i + 1]
    return i + 1

arr = [12, 23, -100, 6, 69, 49, 49, 49, 12, 2037, -8]
print(arr)
r = len(arr)
quick_sort(arr, 0, r - 1)
print(arr)

File: quick_sort.cc
/*
 * Quick sort
 * Nov 24, 2018
 * Divide and conquer..
 */

#include <iostream>
#include <vector>

using namespace std;

int partition(vector<int>& arr, int low, int high) {
  int key = arr[high];

  int pioneer = low;
  int trailer = pioneer - 1;
  while (pioneer < high) {
    if (arr[pioneer] <= key) {
      swap(arr[++trailer], arr[pioneer++]);
    } else
      ++pioneer;
  }

  swap(arr[++trailer], arr[high]);
  return trailer;
}

void quick_sort(vector<int>& arr, int low, int high) {
  if (low < high) {
    int pivot = partition(arr, low, high);
    quick_sort(arr, low, pivot - 1);
    quick_sort(arr, pivot + 1, high);
  }
}

void print_arr(vector<int>& arr) {
  for (auto a : arr) {
    cout << a << " ";
  }
  cout << endl;
}

int main() {
  vector<int> arr{2037, -1, 8, 5, 8, 4, 5, 9, 6, 10, 8};
  print_arr(arr);
  quick_sort(arr, 0, arr.size() - 1);
  print_arr(arr);

  return 0;
}
